docs: add runbook for SWIFT v2 delegated-NIC wedged node mitigation (#6363)
Azure/ARO-HCP: docs/ops/mitigate-swiftv2-wedged-node.md

# Mitigation: SWIFT v2 Delegated-NIC Wedged Node

## Overview

> Azure SWIFT v2 management-cluster nodes can enter a permanent failure state where every pod scheduled to the node fails with `FailedCreatePodSandBox`. The root cause is an Azure platform issue (Cloudnet/RNC) where the delegated FrontendNIC virtual function (VF) is torn down and re-added at the platform level just after CNI attach, leaving the node's network stack broken. The CNI errors are the symptom, not the cause.

Reviewer: Two facts worth putting right here, because they stop people digging in the wrong place.

```suggestion
Azure SWIFT v2 management-cluster nodes can enter a failure state where every pod scheduled to the node fails with `FailedCreatePodSandBox`. The root cause is an Azure platform issue (Cloudnet/RNC) where the delegated FrontendNIC virtual function (VF) is torn down and re-added at the platform level just after CNI attach, leaving the node's network stack broken. The CNI errors are the symptom, not the cause.

Two things to know before you start digging:

- **CNS is not the culprit, so do not spend time there.** On the cases we have analysed, `azure-cns` assigned and released the pod's IP cleanly on every single retry with `ReturnCode:Success`, and the `MultitenantPodNetworkConfig` resolved a secondary IP within about a second of the pod being scheduled. IP pool exhaustion is ruled out too. The fault is in the `azure-vnet` CNI plugin's endpoint-creation step, which consumes the CNS response and programs the interface into the pod netns.
- **`azure-vnet` has no logs in Kusto.** It runs as a host binary invoked by containerd, not as a pod, so nothing it emits is ingested into ServiceLogs. That is exactly why the log-collection script in the appendix exists: if Cloudnet/RNC want evidence, it has to be pulled off the node before you recycle it.
```

The node will not self-heal. The standing mitigation is to **cordon and delete the wedged VMSS instance** so pods reschedule onto healthy nodes and AKS replaces the instance.

## Tracking

- **ICM**: 832382845 (Azure-owned, Cloudnet/RNC root cause)
- **Past occurrences**:
  - `prod uksouth` — node `aks-userswft2-40171262-vmss000004`, ~18h of continuous failure, ~2,584 events (AROSLSRE-1524)

>   - `prod australiaeast` — node `aks-userswft2-40474110-vmss00000f`, ~3 days of continuous failure, ~2,419 events (AROSLSRE-1563 / AROSLSRE-1585)

Reviewer: Missing the most recent one.

```suggestion
  - `prod australiaeast` — node `aks-userswft2-40474110-vmss00000f`, ~3 days of continuous failure, ~2,419 events (AROSLSRE-1563 / AROSLSRE-1585)
  - `prod centralindia`, node `aks-userswft3-38105468-vmss000005`, ongoing since 2026-08-01, ~1,074 events, took out three consecutive prod e2e runs on 2026-08-04 (AROSLSRE-1717)
```

## Symptoms

All of these will be present on the affected node:

| Signal | Detail |
|---|---|
| **Event** | `FailedCreatePodSandBox` at ~15s retry cadence, thousands of occurrences |
| **Error messages** | `route ip+net: no such network interface`, `mtpnc is not ready`, `dhcp discover timed out` |
| **Blast radius** | 100% of pods scheduled to the node fail; hosted-cluster router replicas are especially impacted → `KASLoadBalancerNotReachable` → cluster-create timeouts |

> | **Isolation** | Only one node is affected; other nodes in the same VMSS pool are healthy |

Reviewer: This one I would soften. It reads as a discriminator, and if someone uses it to rule the issue out they will get it wrong.

In AROSLSRE-1717 five HCP namespaces hit the identical CNI errors inside the same six-minute window, across more than one node, and four of them recovered. So concentration on a single node is the common case, not a reliable test.

```suggestion
| **Isolation** | Failures usually concentrate on one node, but a short cluster-wide burst across several nodes is possible (see AROSLSRE-1717). Confirm with per-node event counts rather than assuming a single node |
```

| **Duration** | Does not self-heal; persists for days until the node is deleted |

## Diagnosis

### 1. Identify the wedged node via Kusto (hcp-prod-{region})

```kql
// Look for sustained FailedCreatePodSandBox on a single node
ContainerEvent
| where Reason == "FailedCreatePodSandBox"
| where Message has "no such network interface" or Message has "mtpnc is not ready" or Message has "dhcp discover timed out"
| summarize EventCount = count(), FirstSeen = min(TimeGenerated), LastSeen = max(TimeGenerated) by NodeName

> | order by EventCount desc

Reviewer: `ContainerEvent` does not exist in our Kusto. Kubernetes events land in `ServiceLogs.kubernetesEvents`, and the columns are lowercase (`timestamp`, `host`, `reason`, `message`), see [docs/ai/kusto-debugging.md](https://github.com/Azure/ARO-HCP/blob/main/docs/ai/kusto-debugging.md). As written this query fails with table-not-found, which is the worst thing that can happen in a runbook someone opens mid-incident.

Same problem in the other two KQL blocks below.

```suggestion
// Look for sustained FailedCreatePodSandBox, per node.
// Run against the region's cluster, database ServiceLogs.
kubernetesEvents
| where timestamp > ago(7d)
| where reason == "FailedCreatePodSandBox"
| where message has "no such network interface"
     or message has "mtpnc is not ready"
     or message has "dhcp discover"
| summarize events=count(), nsAffected=dcount(eventNamespace), podsAffected=dcount(objectName),
            firstSeen=min(timestamp), lastSeen=max(timestamp)
  by cluster, host
| order by events desc
```

```

### 2. Distinguish hard-wedge from transient flap

This is the critical decision gate. Not every occurrence of these errors requires node deletion. The same Azure fault (ICM 832382845) has two modes:

| Mode | Behavior | Action |
|---|---|---|
| **Hard-wedge** | Zero successful sandbox creations, continuous error storm (2500-5300+ events), VF gone, never recovers | Cordon + delete node |
| **Transient flap** | Bounded error bursts with successful DHCP sends interleaved, self-heals within minutes to hours | Alert only, do not recycle |

```kql
// Flap vs. wedge: check for interleaved successes on the suspect node.
// If SuccessCount > 0, the node is flapping (self-healing). Do NOT delete.

> // If SuccessCount == 0 for >= 30-60 min of continuous errors, it is wedged.
> let suspect_node = "<node-name>";
> let lookback = 2h;
> ContainerEvent
> | where TimeGenerated > ago(lookback)
> | where NodeName == suspect_node
> | where Reason in ("FailedCreatePodSandBox", "Created", "Started")
> | summarize
>     FailCount = countif(Reason == "FailedCreatePodSandBox"
>                         and (Message has "no such network interface"
>                              or Message has "mtpnc is not ready"
>                              or Message has "dhcp discover timed out")),
>     SuccessCount = countif(Reason in ("Created", "Started")),
>     FirstFail = minif(TimeGenerated, Reason == "FailedCreatePodSandBox"),
>     LastFail = maxif(TimeGenerated, Reason == "FailedCreatePodSandBox"),
>     LastSuccess = maxif(TimeGenerated, Reason in ("Created", "Started"))
>     by NodeName
> | extend FailDuration = LastFail - FirstFail,
>          TimeSinceLastSuccess = LastFail - LastSuccess
> | project NodeName, FailCount, SuccessCount, FailDuration,
>           TimeSinceLastSuccess, FirstFail, LastFail, LastSuccess

Reviewer: Same table and column fix here. I also swapped `NodeName` for `host` and dropped the `let` for the cluster scope, since you almost always want to pin `cluster` too.

```suggestion
let suspect_node = "<node-name>";
let suspect_cluster = "<mgmt-cluster-name>";
let lookback = 2h;
kubernetesEvents
| where timestamp > ago(lookback)
| where cluster == suspect_cluster and host == suspect_node
| where reason in ("FailedCreatePodSandBox", "Created", "Started")
| summarize
    FailCount = countif(reason == "FailedCreatePodSandBox"
                        and (message has "no such network interface"
                             or message has "mtpnc is not ready"
                             or message has "dhcp discover")),
    SuccessCount = countif(reason in ("Created", "Started")),
    FirstFail = minif(timestamp, reason == "FailedCreatePodSandBox"),
    LastFail = maxif(timestamp, reason == "FailedCreatePodSandBox"),
    LastSuccess = maxif(timestamp, reason in ("Created", "Started"))
    by host
| extend FailDuration = LastFail - FirstFail,
         TimeSinceLastSuccess = LastFail - LastSuccess
| project host, FailCount, SuccessCount, FailDuration,
          TimeSinceLastSuccess, FirstFail, LastFail, LastSuccess
```

```bash
# Target only the wedged node
NODES_FILTER='<vmss-instance-id>' ./collect-swiftv2-node-logs.sh
```

The script is read-only: it creates temporary busybox debug pods, pulls `/host/var/log/azure-vnet*` and `/host/var/run/azure-cns` state, tars them, and deletes only the debug pods it created. See the full script in the appendix below or ask the on-call SRE for the latest version.

### Step 2: Cordon the node

```bash
kubectl cordon <node-name>
```

This prevents new pods from being scheduled to the node while you verify and delete.

### Step 3: Delete the VMSS instance

```bash
# Get the VMSS name and instance ID from the node's provider ID
PROVIDER_ID=$(kubectl get node <node-name> -o jsonpath='{.spec.providerID}')
echo "${PROVIDER_ID}"
# Format: azure:///subscriptions/.../virtualMachineScaleSets/<vmss-name>/virtualMachines/<instance-id>

# Delete via az cli (the VMSS will automatically provision a replacement)
az vmss delete-instances \
  --resource-group <node-resource-group> \
  --name <vmss-name> \
  --instance-ids <instance-id>
```

Alternatively, if you only have `kubectl` access:

```bash
kubectl delete node <node-name>
```

> > **Note**: Deleting the node object from Kubernetes does not delete the underlying VMSS instance. For a full cleanup, use `az vmss delete-instances`. The AKS cluster autoscaler or node-pool reconciler will provision a replacement.

Reviewer: Worth adding a durability caveat here. AROSLSRE-1717 shows the faulty-node role migrating across VMSS instances over weeks, with clean handoffs (one instance stops failing the same minute the next one starts). Deleting the instance fixes the incident in front of you but has not held.

```suggestion
> **Note**: Deleting the node object from Kubernetes does not delete the underlying VMSS instance. For a full cleanup, use `az vmss delete-instances`. The AKS cluster autoscaler or node-pool reconciler will provision a replacement.

> **This mitigation is not durable.** The fault has been observed moving to a different instance in the same pool, sometimes within minutes of the previous one being recycled (AROSLSRE-1717). Recycling clears the immediate incident, it does not fix the underlying platform issue. Update ICM 832382845 on every recurrence so Azure has the pattern.
```

### Step 4: Verify recovery

```bash
# Confirm the node is gone
kubectl get nodes | grep <node-name>

# Confirm pods are rescheduling to healthy nodes
kubectl get pods -A --field-selector status.phase!=Running,status.phase!=Succeeded | head -20

# Confirm no new FailedCreatePodSandBox events on other nodes
kubectl get events --field-selector reason=FailedCreatePodSandBox --sort-by='.lastTimestamp' | tail -10
```

## Post-Mitigation

### 1. Verify no new sandbox failures (Kusto)

```kql
// Run against hcp-prod-{region} after node deletion.
// Expect 0 results for the deleted node, and no new nodes exhibiting the same pattern.
ContainerEvent
| where TimeGenerated > ago(30m)
| where Reason == "FailedCreatePodSandBox"
| where Message has "no such network interface"
     or Message has "mtpnc is not ready"
     or Message has "dhcp discover timed out"
| summarize EventCount = count(), LastSeen = max(TimeGenerated) by NodeName

> | order by EventCount desc

Reviewer: Same fix for the post-mitigation check.

```suggestion
// Run against the region's cluster, database ServiceLogs, after node deletion.
// Expect 0 results for the deleted node, and no new nodes showing the same pattern.
kubernetesEvents
| where timestamp > ago(30m)
| where reason == "FailedCreatePodSandBox"
| where message has "no such network interface"
     or message has "mtpnc is not ready"
     or message has "dhcp discover"
| summarize events=count(), lastSeen=max(timestamp) by cluster, host
| order by events desc
```

| Condition | This issue (SWIFT v2 VF churn) | General node NotReady |
|---|---|---|
| Node status | Usually `Ready` (kubelet is fine) | `NotReady` |
| Failure mode | Only pod sandbox creation fails | All node operations impacted |
| Error signature | `no such network interface` / `mtpnc` / `dhcp discover` | Varies |
| AKS auto-repair triggers? | **No** -- node is `Ready`, so auto-repair does not engage | Yes -- AKS escalates reboot, reimage, redeploy, delete |
| Self-healing | Never (manual deletion required) | Often recovers via auto-repair |
| Node pool | `userswft2` / `userswft3` (SWIFT v2 pools) | Any pool |

## Appendix: Node Healer Design Takeaways

From the investigation of the first observed self-heal (AROSLSRE-1642/1643), these criteria distinguish actionable wedges from transient flaps. Any future automated node healer should implement these rules:

1. **Key on concurrent success on the node** -- do not wait passively for self-heal.
2. **0 successful sandbox creations for >= 30-60 min of continuous route errors** → wedge → cordon/drain/delete.
3. **Any successes interleaved with the failures** → flap → alert only, do not recycle.

> 4. Both conditions are queryable in `hcp-prod-{region}` ServiceLogs.

Reviewer: This is no longer hypothetical. The healer shipped in [#6284](https://github.com/Azure/ARO-HCP/pull/6284) and is being turned on by default in [#6394](https://github.com/Azure/ARO-HCP/pull/6394) (AROSLSRE-1588), and it implements exactly these rules.

That also makes diagnosis cheaper than hand-writing Kusto, so it is worth pointing at from step 1 as well.

````suggestion
## Appendix: Automated Detection (node-health controller)

The criteria below came out of the first observed self-heal (AROSLSRE-1642/1643) and are now implemented by the `node-health` controller in mgmt-agent (AROSLSRE-1588):

1. **Key on concurrent success on the node**, do not wait passively for self-heal.
2. **0 successful sandbox creations for >= 30-60 min of continuous route errors** → wedge → cordon/drain/delete.
3. **Any successes interleaved with the failures** → flap → alert only, do not recycle.

Where it has rolled out, a wedged node carries:

- label `node-health.aro-hcp.azure.com/status=wedged`
- annotations `node-health.aro-hcp.azure.com/detector`, `/reason`, `/signature`, `/observed-at`
- metrics `nodehealth_node_wedged`, `nodehealth_detections_total`, `nodehealth_label_actions_total`

So the quickest first pass at diagnosis is:

```bash
kubectl get nodes -l node-health.aro-hcp.azure.com/status=wedged
kubectl get node <node-name> -o jsonpath='{.metadata.annotations}' | jq
```

The controller only labels and annotates, it does not cordon, drain or delete. The mitigation steps in this runbook are still manual. Fall back to the Kusto queries above if the label is not present, for example on a cluster that has not picked up the controller yet.
````

```bash
# Check if the system pool is already in a wedged/Updating state
# (the self-healer keys on provisioningState=Updating + NRP-KVS error storm)
az aks nodepool show \
  --resource-group <resource-group> \
  --cluster-name <cluster-name> \
  --name system \
  --query '{state: provisioningState, count: count}' -o table

# Check for NRP-KVS (NetworkingInternalOperationError) events in the last 6h
# If the self-healer's guards are passing, it will handle the remediation
```

> **Note**: The `recreate-system-pool` self-healer targets NRP-KVS wedge scenarios specifically; it may not cover the SWIFT v2 VF churn issue if the system pool's `provisioningState` is `Succeeded` (not `Updating`). If the self-healer is not applicable, proceed with manual mitigation.

### 6. Confirm router impact (if cluster-creates are timing out)

```bash
# Check if router replicas are landing on the wedged node
kubectl get pods -A -o wide | grep router | grep <node-name>
```

## Prerequisites

> - JIT access to the affected management-cluster AKS (include this runbook link and ICM 832382845 in the justification)

Reviewer 2: what level of access for the JIT?

- `kubectl` configured against the management AKS
- (Optional) The ACN log-collection script if Cloudnet/RNC wants on-node evidence before recycling

## Mitigation Procedure

### Step 1: (Optional) Collect ACN logs for evidence preservation

Run the log-collection script **before** deleting the node, in case the Azure networking team (Cloudnet/RNC) needs on-node state for root-cause analysis.

```bash
# Target only the wedged node

> NODES_FILTER='<vmss-instance-id>' ./collect-swiftv2-node-logs.sh

Reviewer 3: Step 1 invokes ./collect-swiftv2-node-logs.sh but it only exists inline in the appendix. Commit it under docs/ops/ or hack so it's runnable

```

The script is read-only: it creates temporary busybox debug pods, pulls `/host/var/log/azure-vnet*` and `/host/var/run/azure-cns` state, tars them, and deletes only the debug pods it created. See the full script in the appendix below or ask the on-call SRE for the latest version.

### Step 2: Cordon the node

```bash
kubectl cordon <node-name>
```

This prevents new pods from being scheduled to the node while you verify and delete.

> ### Step 3: Delete the VMSS instance

Reviewer 3: Deleting the node, deletes the vmss instance.

```bash
# Get the VMSS name and instance ID from the node's provider ID
PROVIDER_ID=$(kubectl get node <node-name> -o jsonpath='{.spec.providerID}')
echo "${PROVIDER_ID}"

> # Format: azure:///subscriptions/.../virtualMachineScaleSets/<vmss-name>/virtualMachines/<instance-id>

Reviewer 2: Let's fully expand this format as it has useful metadata (such as node-resource-group), which is useful for the step immediately after

```suggestion
# Format: azure:///subscriptions/<subscription-id>/resourceGroups/<node-resource-group>/providers/Microsoft.Compute/virtualMachineScaleSets//virtualMachineScaleSets/<vmss-name>/virtualMachines/<instance-id>
```

# Delete via az cli (the VMSS will automatically provision a replacement)
az vmss delete-instances \
  --resource-group <node-resource-group> \
  --name <vmss-name> \
  --instance-ids <instance-id>
```

Alternatively, if you only have `kubectl` access:

```bash
kubectl delete node <node-name>
```

> > **Note**: Deleting the node object from Kubernetes does not delete the underlying VMSS instance. For a full cleanup, use `az vmss delete-instances`. The AKS cluster autoscaler or node-pool reconciler will provision a replacement.

Reviewer 3: >Deleting the node object from Kubernetes does not delete 

We saw that deleting the node deletes the vmss instance on Tuesday.

## Tracking

- **ICM**: 832382845 (Azure-owned, Cloudnet/RNC root cause)
- **Past occurrences**:
  - `prod uksouth` — node `aks-userswft2-40171262-vmss000004`, ~18h of continuous failure, ~2,584 events (AROSLSRE-1524)
  - `prod australiaeast` — node `aks-userswft2-40474110-vmss00000f`, ~3 days of continuous failure, ~2,419 events (AROSLSRE-1563 / AROSLSRE-1585)
  - `prod centralindia` — node `aks-userswft3-38105468-vmss000005`, intermittent flapping since 2026-08-01, ~1,074 events, took out three consecutive prod e2e runs on 2026-08-04; node deleted 2026-08-04 via `kubectl delete node` (AROSLSRE-1717 / AROSLSRE-1723)

## Symptoms

All of these will be present on the affected node:

| Signal | Detail |
|---|---|
| **Event** | `FailedCreatePodSandBox` at ~15s retry cadence, thousands of occurrences |
| **Error messages** | `route ip+net: no such network interface`, `mtpnc is not ready`, `dhcp discover timed out` |

> | **Blast radius** | 100% of pods scheduled to the node fail; hosted-cluster router replicas are especially impacted → `KASLoadBalancerNotReachable` → cluster-create timeouts |

Reviewer: Same point as the test-pod step, and it is the reason that step looks safe. Going by the `SecondaryEndpointClient` signature and the pods I captured on the uksouth node, what fails is the pods that need a delegated NIC, which in practice is the hosted-control-plane workload. I would not claim 100% of pods, because that is exactly what makes a plain test pod look like a valid probe.

```suggestion
| **Blast radius** | Every pod needing a delegated NIC fails on that node, which in practice is the whole hosted-control-plane workload scheduled there; hosted-cluster router replicas are especially impacted → `KASLoadBalancerNotReachable` → cluster-create timeouts |
```

> | **Blast radius** | 100% of pods scheduled to the node fail; hosted-cluster router replicas are especially impacted → `KASLoadBalancerNotReachable` → cluster-create timeouts |

Author: Good catch — will fix in next push. Changing to:
> Every pod needing a delegated NIC fails on that node, which in practice is the whole hosted-control-plane workload scheduled there; hosted-cluster router replicas are especially impacted → `KASLoadBalancerNotReachable` → cluster-create timeouts

> - Any successes interleaved with the failures → **flap** → alert only, do not recycle
>
> Both are queryable in `hcp-prod-{region}` ServiceLogs.

### 3. Confirm the node is still failing

Kubernetes events have a short TTL (~1h). If the wedge is intermittent (flapping variant), kubectl events may show nothing even though the fault is still latent.

**kubectl (if events are recent):**

```bash
kubectl get events -A --field-selector reason=FailedCreatePodSandBox | grep <node-name> | tail -20
```

**Kusto (reliable regardless of event TTL):**

```kql
kubernetesEvents
| where timestamp > ago(4h)
| where host == "<node-name>"
| where reason == "FailedCreatePodSandBox"
| order by timestamp desc
| take 20
```

**Test pod (confirms active vs. latent fault):**

If Kusto shows recent failures but kubectl events are empty, schedule a test pod directly on the suspect node to check whether the NIC is currently functional:

```bash
kubectl run nic-test --image=mcr.microsoft.com/cbl-mariner/base/core:2.0 \
    --overrides='{"spec":{"nodeName":"<node-name>"}}' \
    --command -- sleep 300
kubectl get pod nic-test -w
```

> - If the pod reaches `Running`: the NIC is not actively wedged right now. The fault is intermittent and will likely recur under scheduling pressure (e.g. during e2e runs). Proceed with the flap-vs-wedge query (Step 2) to decide whether to delete.

Reviewer: This is the one I would change.

The error we actually see is `SecondaryEndpointClient Error: route ip+net: no such network interface`. `SecondaryEndpointClient` is the delegated-NIC path in azure-vnet, so it only runs for pods that get an MTPNC. `kubectl run nic-test` in whatever namespace the operator happens to be in has no `PodNetworkInstance` and no delegated NIC, so it takes the ordinary endpoint path and never reaches the code that is broken. On the uksouth node I captured, every failing pod was in an `ocm-arohcpprod-*` namespace.

So a `Running` nic-test does not tell you the node is healthy, it tells you the primary CNI path works. Right now the step reads it as "the NIC is not actively wedged", and someone using that to rule the issue out will leave a genuinely wedged node in the pool.

I would keep the pod as a positive signal only (if it fails you are definitely wedged) and make the pass case explicitly inconclusive:

```suggestion
- If the pod reaches `Running`: this is inconclusive, not a clean bill of health. The probe pod has no delegated NIC, so it exercises the ordinary CNI path and not the `SecondaryEndpointClient` path that fails here. Decide from the flap-vs-wedge query (Step 2), not from this pod.
```

> - If the pod reaches `Running`: the NIC is not actively wedged right now. The fault is intermittent and will likely recur under scheduling pressure (e.g. during e2e runs). Proceed with the flap-vs-wedge query (Step 2) to decide whether to delete.

Author: Agreed — the probe pod has no delegated NIC so it exercises the ordinary CNI path, not the `SecondaryEndpointClient` path. Will update the Running case to say this is inconclusive rather than a clean bill of health.

```bash
# Check if the system pool is already in a wedged/Updating state
# (the self-healer keys on provisioningState=Updating + NRP-KVS error storm)
az aks nodepool show \
  --resource-group <resource-group> \
  --cluster-name <cluster-name> \
  --name system \
  --query '{state: provisioningState, count: count}' -o table

# Check for NRP-KVS (NetworkingInternalOperationError) events in the last 6h
# If the self-healer's guards are passing, it will handle the remediation
```

> **Note**: The `recreate-system-pool` self-healer targets NRP-KVS wedge scenarios specifically; it may not cover the SWIFT v2 VF churn issue if the system pool's `provisioningState` is `Succeeded` (not `Updating`). If the self-healer is not applicable, proceed with manual mitigation.

### 6. Confirm router impact (if cluster-creates are timing out)

```bash
# Check if router replicas are landing on the wedged node
kubectl get pods -A -o wide | grep router | grep <node-name>
```

## Prerequisites

- JIT access to the affected management-cluster AKS (include this runbook link and ICM 832382845 in the justification)

> - `kubectl` configured against the management AKS

Author: Done in commit b3493f27 — the providerID format line now reads:
```
azure:///subscriptions/<subscription-id>/resourceGroups/<node-resource-group>/providers/Microsoft.Compute/virtualMachineScaleSets/<vmss-name>/virtualMachines/<instance-id>
```
This shows all the metadata needed for the `az vmss delete-instances` step that follows.

> **Note on the centralindia incident (AROSLSRE-1723)**: Both **AKS RBAC Cluster Admin** and **Change Safety Contributor** were granted via JIT. Neither was sufficient for `az vmss delete-instances` on this subscription. The mitigation succeeded via `kubectl delete node` instead — the node object was removed immediately, AKS deprovisioned the VMSS instance, and a replacement node was provisioned ~15 minutes later (reaching Ready ~40 seconds after creation).

## Mitigation Procedure

### Step 1: (Optional) Collect ACN logs for evidence preservation

Run the log-collection script **before** deleting the node, in case the Azure networking team (Cloudnet/RNC) needs on-node state for root-cause analysis. The script lives alongside this runbook at [`collect-swiftv2-node-logs.sh`](collect-swiftv2-node-logs.sh).

```bash
# Target only the wedged node (run from this directory: docs/ops/)

> NODES_FILTER='<vmss-instance-id>' ./collect-swiftv2-node-logs.sh

Reviewer: `NODES_FILTER` is used as a `grep -E` pattern against node names, and the script's own header shows `NODES_FILTER='vmss000004'`. An instance ID is a bare number, so as a regex it matches any node with that digit anywhere in its name and you collect from more nodes than you meant.

```suggestion
NODES_FILTER='<node-name>' ./collect-swiftv2-node-logs.sh
```

> NODES_FILTER='<vmss-instance-id>' ./collect-swiftv2-node-logs.sh

Author: Will fix — changing to `NODES_FILTER='<node-name>'` since the filter is used as a `grep -E` pattern against node names.

```

The script is read-only: it creates temporary busybox debug pods, pulls `/host/var/log/azure-vnet*` and `/host/var/run/azure-cns` state, tars them, and deletes only the debug pods it created.

### Step 2: Cordon the node

```bash

> kubectl cordon <node-name>

Reviewer: Worth a line here about what Step 3 then does to the pods still on the node.

Cordon only stops new scheduling. `kubectl delete node` makes the pod garbage collector force-delete everything bound to that node, and that path does not go through the Eviction API, so PDBs are not consulted and there is no graceful termination. On a mgmt cluster those are customer control-plane pods, and the ones that started before the wedge began are still serving fine.

Either put a drain between cordon and delete:

```bash
kubectl drain <node-name> --ignore-daemonsets --delete-emptydir-data --timeout=5m
```

or say explicitly that we skip it and why. I do not mind which one, but as written the reader cannot tell it was a decision.

> kubectl cordon <node-name>

Author: Good call — will add an explicit `kubectl drain` step between cordon and delete, with a note that skipping drain means PDBs are not consulted and existing healthy pods get force-deleted.
